### PR TITLE
Encode string values right after the generation process

### LINKED ISSUE #40: extra single quote in URL for datetime
sample URL part: `FlightDate gt datetime''7851-03-24T05:08:23''`

### PR DESCRIPTION
**Pull request by lubosmj**
Closes #40

### CODE REVIEW

File: odfuzz/config.py [hunk 1]
```diff
@@ -2,7 +2,21 @@
 
 import os
 
-from odfuzz.constants import ENV_ODFUZZ_CERTIFICATE_PATH, DEFAULT_ASYNC_REQUESTS_NUM, DEFAULT_SAP_CLIENT, DEFAULT_DATA_FORMAT, DEFAULT_URLS_PER_PROPERTY, ENV_SAP_CLIENT, ENV_DATA_FORMAT, ENV_URLS_PER_PROPERTY, ENV_ASYNC_REQUESTS_NUM
+from odfuzz.constants import (
```
Comment (phanak-sap) [inline odfuzz/config.py:5]: even tho it is definitely better, it is good practice to not mix PR containing fix or feature with changes that are formatting only. This should ideally be different PR.

File: odfuzz/entities.py [hunk 2]
```diff
@@ -109,10 +109,6 @@ def __init__(self, metadata, restrictions):
         self._metadata_string = metadata
         self._restrictions = restrictions
 
-        Config.init()
```
Comment (phanak-sap) [inline odfuzz/entities.py:112]: what is the rationale for this removal? Also does not seems to be relevant to the bug itself.

File: odfuzz/config.py [hunk 3]
```diff
@@ -2,7 +2,21 @@
...
+    DEFAULT_SAP_CLIENT,
+    DEFAULT_URLS_PER_PROPERTY,
+    ENV_ASYNC_REQUESTS_NUM,
+    ENV_DATA_FORMAT,
+    ENV_ENCODE_STRING,
+    ENV_ODFUZZ_CERTIFICATE_PATH,
+    ENV_SAP_CLIENT,
+    ENV_URLS_PER_PROPERTY,
+)
+from odfuzz.utils import encode_double, encode_string
```
Comment (phanak-sap) [inline odfuzz/config.py:18]: I don't understand why this should be in Config and later called this way from generators. 
seems unnecessary complicated, when usually the config just stores the values grabbed from CLI or env variable and thats all he does.. all logic based on the value of such variable in place where is actually used.

File: requirements.txt [hunk 4]
```diff
@@ -1,4 +1,4 @@
-gevent>=1.2.2
+gevent~=1.5.0
```
Comment (phanak-sap) [inline requirements.txt:1]: please create separate PR or issue for that. This PR should really focus on bugfix and not update dependencies. It is not good idea to move too many things in one block.

File: tests/databases/test_pymongo.py [hunk 5]
```diff
@@ -140,7 +142,7 @@ def test_database_total_score_eleven(data_single_filter_logical_company_code, da
     assert total_score == 11
 
 
-'''
+@pytest.mark.skip(reason="Cannot mock random sampling")
```
Comment (phanak-sap) [inline tests/databases/test_pymongo.py:145]: Not sure why this is skipped. If the test is useless, perhaps delete it and do this in another PR. Seems not relevant to the topic of this bugfix PR.

File: odfuzz/config.py [hunk 6]
```diff
@@ -45,7 +84,10 @@ def async_requests_num(self):
...
 class Config:
+    fuzzer = None
+    dispatcher = None
+
     @staticmethod
     def init():
         Config.fuzzer = FuzzerConfig()
-        Config.dispatcher = DispatcherConfig()
\ No newline at end of file
+        Config.dispatcher = DispatcherConfig()
```
Comment (phanak-sap) [inline odfuzz/config.py:93]: is this just different newline? imho should not be present in this PR. Leave pls re-formatting away

File: requirements.txt (hunk 4 above)
Comment (lubosmj) [reply requirements.txt:1]: ~= means that it is pinned to the latest version of the package, greater or equal to 1.5.0, but still in 1.5.*. I cannot test the code without this modification because it is giving me errors. So, I think this change should be included too.

File: tests/databases/test_pymongo.py (hunk 5 above)
Comment (lubosmj) [reply tests/databases/test_pymongo.py:145]: That is why I split changes into multiple commits -- to not lose track of the changes. For instance, I still need to update docs. Every single improvement to the existing codebase should be welcomed.

File: odfuzz/config.py (hunk 6 above)
Comment (phanak-sap) [reply odfuzz/config.py:93]: ok.

File: requirements.txt (hunk 4 above)
Comment (phanak-sap) [reply requirements.txt:1]: created #48 for this.

File: odfuzz/config.py [hunk 7]
```diff
@@ -1,8 +1,24 @@
 """This module contains classes for fetching and parsing basic configurations."""
 
+import importlib
```
Comment (phanak-sap) [inline odfuzz/config.py:3]: why is this change needed?

File: requirements.txt (hunk 4 above)
Comment (phanak-sap) [reply requirements.txt:1]: > I cannot test the code without this modification because it is giving me errors. So, I think this change should be included too.

Not at all. So far in new clean isolated environment this does not produce any error with usage relevant to gevenet since it istall latest 20.5.0. I am seeing that for more than half a year on internal servers. The error may really be just your local environment. 

Fix that is essentially "different string replace" should imho not depend on upgrade of "coroutine networking library" - it is a problem of what is produced, not where it is used.

If it would be really problem I would still rather discuss and merge #48 first and have bugfix PR isolated of dependency change PR.

File: tests/databases/test_pymongo.py (hunk 5 above)
Comment (phanak-sap) [reply tests/databases/test_pymongo.py:145]: it is true and it is welcomed. But it is no problem to have smaller, topic-focused PRs. And update  relevant docs in another PR when the bug is fixed even tho for some time there would be no mention of new configuration in documentation after merge. Issue with label documentation is enough and then have review of different PR.

File: odfuzz/config.py (hunk 7 above)
Comment (lubosmj) [reply odfuzz/config.py:3]: This module `importlib` is used for importing modules which are out of the scope during the runtime. I am using it to get a function object from the configuration specified by a user. The configuration looks like this: https://github.com/SAP/odfuzz/pull/47/files#diff-31a9af601fb02ccdbe2b445fc0628e6eR53-R54

File: tests/databases/test_pymongo.py (hunk 5 above)
Comment (lubosmj) [reply tests/databases/test_pymongo.py:145]: You are totally right, I opened the new PR https://github.com/SAP/odfuzz/pull/49 for this.

File: odfuzz/odfuzz.py [hunk 8]
```diff
@@ -8,7 +8,6 @@
 import gevent
 
 from datetime import datetime
-from functools import partial
```
Comment (phanak-sap) [inline odfuzz/odfuzz.py:11]: why is this deleted? Isnt it relevant to the gevent problem?

File: odfuzz/config.py (hunk 7 above)
Comment (phanak-sap) [reply odfuzz/config.py:3]: But is it really necessary? I mean, " importing modules which are out of the scope during the runtime." looks like a problem on its own and the importlib is just a bandage slapped on it. 

- This will most probably break any static analysis relevant to inner dependencies (e.g. pydeps), which we need to emply anyway to better separate modules, instead of cyclic dependencies.
- If adding configuration in current state of codebase is such a problem, I would prefer to just directly fix the encoded without any configuration and open a new issue for such. 
- in my opinion, it would be much simpler and direct to just use true constant for DEFAULT_STRING_ENCODER/DECODER, instead of passing function back and forth and solve the import and probably gevent related problems.  

I would like to move in a way that the runtime stuff is one block, isolated away from fuzzer, which should be perhaps even a true isolated python library, for generating URL from odata V2/V4 only - essencially entities, generators, mutators.  It will be more testable and not inertwined with stuff like networking, greenlets or database. 

If you agree with this vision that could be better, then is obviously a bad call to add something like Config.fuzzer.decode_string(search_option  and call it in fuzzer.py, while in effect it is about generators and mutators and the string encoder change propagated there as well.

File: requirements.txt (hunk 4 above)
Comment (phanak-sap) [reply requirements.txt:1]: created PR #52 for this change, after all comments in #48. Please comment there before merging. #52 should be merged before this one.

File: README.md [hunk 9]
```diff
@@ -158,7 +164,7 @@ Taking this into account, you have to set the fuzzer's output directories to /im
 $ odfuzz <SERVICE_URL> -l logs_directory -s stats_directory -r restrictions/basic.yaml -a -f -p
 ```
 
-The option **-a** enables fuzzer to send asynchronous requests. A default number of the asynchronous requests can be changed. To do so, navigate to the file *config/fuzzer/config.yaml* and modify value *async_requests_num*. Notice that some services do not support more than 10 asynchronous requests at the same time.
+The option **-a** enables the fuzzer to send concurrently multiple requests. The default number of the asynchronous requests can be changed. To do so, use the environment variable `ODFUZZ_ASYNC_REQUESTS_NUM`. Notice that some services do not support handling of more than 10 asynchronous requests at the same time.
```
Comment (phanak-sap) [inline README.md:167]: The part about "not support handling more than 10 asynch" is I believe actually not true.

The number 10 is default value, based on idea of not overloading our sap test servers with multiple odfuzz instances in the wild. However I think user can provide any number in the env variable `export ODFUZZ_ASYNC_REQUESTS_NUM=10` and its up to gevent and target server to handle it properly.

**Comment by phanak-sap** [comment]
I like how this is becoming smaller and more contained due separation of different stuff to different PRs :) Keep up the good work.

File: README.md [hunk 10]
```diff
@@ -87,9 +87,15 @@ export ODFUZZ_ASYNC_REQUESTS_NUM=10
 
 File path where the HTTPS certificate is stored if the service is requiring it.
 ```
-export ENV_ODFUZZ_CERTIFICATE_PATH=./cert.crt
+export ODFUZZ_CERTIFICATE_PATH=./cert.crt
```
Comment (phanak-sap) [inline README.md:90]: thx for correcting, obviously my copypaste error :)

File: README.md [hunk 11]
```diff
@@ -87,9 +87,15 @@ export ODFUZZ_ASYNC_REQUESTS_NUM=10
...
 ```
 
+A default encoder and decoder for generated strings. The encoder is used for encoding strings which may contain special characters. The decoder is used then for decoding the string back to its original representation.
```
Comment (phanak-sap) [inline README.md:93]: should be better to use "generated url parts" instead of "generated strings"? the latter could have many possible meaning and even with this documentation I would not be sure what the configuration affects, if I would not spend such time in PR review.

similarly the env. variable should be ODFUZZ_URL_PART_ENCODER or something similar, that would point just by its name what it really affects.

File: odfuzz/config.py (hunk 7 above)
Comment (phanak-sap) [reply odfuzz/config.py:3]: seems that we get anywehere in this  long question than "it just stays there" :)

File: README.md (hunk 11 above)
Comment (phanak-sap) [reply README.md:93]: or correct if I am wrong.. I noticed that the encoder is not used for Edm.DateTime but the PR is marked as closes #40.
It is intentional?

It may be worth to write to documentation list of odata V2 types that are affected by this configuration option.

File: tests/test_utils.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
+from odfuzz.utils import encode_string, decode_string
```
Comment (phanak-sap) [inline tests/test_utils.py:1]: This is a good unit test for the utils method, that's OK.

But ideally we should have tests for all EDM data types with the current encoder/decoder on and off. If I am correct, the issue #40 is not yet reproduced by any existing tests and current generator and mutator tests don't handle the encoding even that both classes has been changed in this PR.

**Comment by phanak-sap** [comment]
If the options for the configuration set by env values are in effect "encoder on" and "encoder off" and seems there is no reason for third option to be added, should this be a simple boolean flag instead of method name in module? 

The if statement for this value would be just in the utils module, no need for importlib usage, just direct value read instead of setting up function inside configuration that is passed trough the entire project.

Simple, easier to understand, write tests against, log and debug. 

I don't still understand why the more complex solution is applied. Even if there would be rationale for third option, value passing instead of function passing seems more maintainable to me.

File: README.md (hunk 11 above)
Comment (lubosmj) [reply README.md:93]: In the past, ODfuzz had been encoding every single generated **part** of a **query** while building an URL. In this commit, ODfuzz encodes only a specific subset of generated **strings** coming straight from the generators. These strings shape one **query** when combined together.

The encoder is not used for Edm.DateTime because it is not necessary to use it there. What do we want to encode in Edm.DateTime? During generation, there should not be present any special characters because the OData protocol forbids that; on the contrary, Edm.String may contain multiple UTF-8 characters which must be encoded.

I can add a list of all affected Edm types to README.

File: odfuzz/config.py (hunk 7 above)
Comment (lubosmj) [reply odfuzz/config.py:3]: The idea behind this approach was to load any encoder/decoder which will properly serve for encoding and decoding. What if a user wants to define his own decoder? The module config.py should be used to initialize the configuration for ODfuzz. As a part of that, it just transforms string value defined by a user to a proper python object that is going to be used during the runtime.

File: tests/test_utils.py (hunk 12 above)
Comment (lubosmj) [reply tests/test_utils.py:1]: The issue may be reproduced by running the integration test and verifying whether a generated URL contains the string in the format `datetime''XXXX-XX-XX''` or not. Otherwise, if you want me to create a new test case specifically for this scenario, I can try to do it.

File: odfuzz/config.py (hunk 7 above)
Comment (phanak-sap) [reply odfuzz/config.py:3]: "What if user wants to define his own X"... this is argument IMHO similar to premature optimization; premature scope creep when there is at the moment not even notable user base for odfuzz as currently is. If we approach any fix with this argument, we wold sunk in configuration to the state that we provide just a "inner platform/own language" for plugins etc. 

My feeling from this is that we started with providing option to user to modify some inner  parts too early and also on wrong end at the moment.  It is a CLI tool after all and it should "just work", since such is the way it is used. It should therefore work well (existing bugs and not covered odata data types or odata protocol versions) before it is necessary to think like "work well in any way user wishes".

We could use similar approach to other areas, that's for sure.. for example the ability to provide own pseudorandom algorithm or some sequence. This would be great in the state of codebase described two three comments above.. library, perhaps even on pip, just for fuzzing URLs based on metadata and other inputs.. and separate project for current CLI runtime with requests and mongo using such library code. 

I think it is a great prototype for how things could be configured, but at the moment, I am more leaning towards no configuration at all, it is a bugfix PR after all. Or if really seems to be necessary to have ability to turn it off for whatever reason I cannot image, just simple boolean value should be enough.

I also think that may be better to apply the the string replace in one place only.. where the url parts are "joining together" in fuzzer.py. Or is there a reason to call in in mutator, generator and fuzzer also? IMHO there should not be problem that generator and mutator creates strings with UTF-8 characters that are not OK for URL parts... encode the url in url-related stuff and just once, right? Or just in mutator and generator and create contract that those modules provide encoded URL parts.

I am looking at this PR and #45 in comparison - quite similar problem, different scope of solving the problem.

File: README.md (hunk 11 above)
Comment (phanak-sap) [reply README.md:93]: README update with list of affected types would be great & enough.

**Comment by lubosmj** [comment]
All comments should be now resolved. Would you mind to review the pull request again?

File: odfuzz/fuzzer.py [hunk 13]
```diff
@@ -102,6 +104,9 @@ def __init__(self, dispatcher, entities, database, output_handler, asynchronous)
             self._queryable_factory = SingleQueryable
             self._dispatch = self._get_single_response
 
+        if Config.fuzzer.use_encoder:
+            self._decode_queries = lambda x: x
```
Comment (phanak-sap) [inline odfuzz/fuzzer.py:108]: Not runinng localy so just just using "Eyeball Mark I" - should this logic be exactly opposite?

I am reading this as "if use_encoder is true ( and that is the deafult value), then use no decoder". I am somehow expecting here the usage of utils.decode_string().

also rename the value pls - "_decode_queries" indicates more boolean value and use in IF statement, similar to _asynchronous or _first_touch. You would not expect such type in _test_num (iditicates int) or _queryable_factory (indicates something.

**Comment by phanak-sap** [comment]
I am still a bit puzzled by usage of function passing instead of value passing, for my taste it is clever yet not really needed. But it current code review it is not something really preventing merging and/or solve in followup PR if really needed.

Please reply on the fuzzer:py108 comment, other than that i looks good, definitely much better then the start of the PR. Also thank you for keeping the commits clean and concise.

**Comment by phanak-sap** [comment]
Also add this bugfix note to changelog.md pls - section "fixes" (it is some new functionality but overall it is not much of a true new feature but more of fix of bad URL generation).

File: odfuzz/fuzzer.py (hunk 13 above)
Comment (lubosmj) [reply odfuzz/fuzzer.py:108]: Oopsie, you are right. This was done unwittingly. I will try to rename the method.

**Comment by lubosmj** [comment]
@phanak-sap, I made the final changes.
